### PR TITLE
Dataset slicing fix

### LINKED ISSUE #906: Slicing datasets fails in some cases
For some reason the `detector_info` table cannot be sliced:

```python
sample = sc.neutron.load('somefile.nxs')
sample.labels['detector_info'].value[Dim.Detector, 0]
```

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-4e0173c6af49> in <module>
----> 1 sample.labels['detector_info'].value[Dim.Detector, 0]

RuntimeError: Dimension size is -1, can't treat 0
```

Is it because there is no data column?

### PR DESCRIPTION
**Pull request by OwenArnold**
fixes #906 

`detector_info` Dataset had no data column so was not able to extract extents. Now looks at `coords`, `attrs` and `masks` in addition to data when extracting extents. Python tests demonstrate working behaviour.

### CODE REVIEW

**Comment by OwenArnold** [comment]
Note that `test_datasetslice.py` is currently disabled in conftest. We should go about fixing up that disabled test.

File: python/bind_slice_methods.h [hunk 1]
```diff
@@ -17,20 +17,50 @@ namespace py = pybind11;
...
+  }
+  return in_extent != out_extent;
+}
+
 // TODO We really need a way to get the dimension labels and extents from
 // Dataset (not using class Dimensions).
 template <class T> auto dim_extent(const T &object, const Dim dim) {
+
   if constexpr (std::is_same_v<T, Dataset> || std::is_same_v<T, DatasetView>) {
     scipp::index extent = -1;
```
Comment (SimonHeybrock) [inline python/bind_slice_methods.h:40]: Doesn't
```cpp
return object.dimensions().at(dim);
```
work to save 30 lines of code?

**Comment by SimonHeybrock** [comment]
Your fix implies that this has always been working correctly on the C++ side?

**Comment by OwenArnold** [comment]
The errors have been coming for the python bindings so yes I believe so https://github.com/scipp/scipp/blob/master/python/bind_slice_methods.h#L83

**Pull request merged**